linera-core/src/client.rs

```diff
@@ -598,57 +600,53 @@ where
-                    errors
-                )
-            }
-        };
-        let signatures: Vec<_> = votes
-            .into_iter()
-            .filter_map(|vote| vote.map(|vote| (vote.validator, vote.signature)))
-            .collect();
-        match action {
+        let (value, round) = match action {
             CommunicateAction::SubmitBlockForConfirmation(proposal) => {
                 let block = proposal.content.block;
-                let round = proposal.content.round;
                 let (executed_block, _) = self.node_client.stage_block_execution(block).await?;
-                let value = HashedValue::from(CertificateValue::ConfirmedBlock { executed_block });
-                let certificate = Certificate::new(value, round, signatures);
-                // Certificate is valid because
-                // * `communicate_with_quorum` ensured a sufficient "weight" of
-                // (non-error) answers were returned by validators.
-                // * each answer is a vote signed by the expected validator.
-                Ok(Some(certificate))
+                (
+                    HashedValue::from(CertificateValue::ConfirmedBlock { executed_block }),
+                    proposal.content.round,
+                )
             }
             CommunicateAction::SubmitBlockForValidation(proposal) => {
                 let BlockAndRound { block, round } = proposal.content;
                 let (executed_block, _) = self.node_client.stage_block_execution(block).await?;
-                let value = HashedValue::new_validated(executed_block);
-                let certificate = Certificate::new(value, round, signatures);
-                Ok(Some(certificate))
+                (HashedValue::new_validated(executed_block), round)
             }
             CommunicateAction::FinalizeBlock(validity_certificate) => {
                 let round = validity_certificate.round;
-                let conf_value = validity_certificate.value.into_confirmed();
-                Ok(Some(Certificate::new(conf_value, round, signatures)))
+                (validity_certificate.value.into_confirmed(), round)
             }
-            CommunicateAction::AdvanceToNextBlockHeight(_) => Ok(None),
-        }
+            CommunicateAction::AdvanceToNextBlockHeight(_) => {
+                return match result {
+                    Ok(_) => Ok(None),
+                    Err(CommunicationError::Trusted(NodeError::InactiveChain(id)))
+                        if id == chain_id =>
+                    {
+                        Ok(None)
+                    }
+                    Err(error) => Err(error)?,
+                };
+            }
+        };
+        let votes = match result? {
+            (Some((votes_hash, votes_round)), votes)
+                if votes_hash == value.hash() && votes_round == round =>
+            {
+                votes
+            }
+            (_, _) => bail!("Unexpected response from validators"),
```

Comment:
`_ => ` also works :)

Reply:
Right; do you prefer that?
I tend to make the catch-all match arm as specific as possible, just so the reader can better see its type.

Reply:
In the (endless) pursue of code standardization, usually the shorter the more acceptable by everyone.